File: csharp/UAParser/UserAgent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace UAParser
{
  /// <summary>
  /// The UserAgent extracted from the user agent string
  /// </summary>
  public class UserAgent
  {
    internal UserAgent() { }
    internal UserAgent(string family, string major, string minor, string patch)
    {
      Family = family;
      Major = major;
      Minor = minor;
      Patch = patch;
    }
    /// <summary>
    /// The family of the user agent
    /// </summary>
    public string Family { get; set; }
    /// <summary>
    /// The major version of the user agent
    /// </summary>
    public string Major { get; set; }
    /// <summary>
    /// The minor version of the user agent
    /// </summary>
    public string Minor { get; set; }
    /// <summary>
    /// The patch version of the user agent
    /// </summary>
    public string Patch { get; set; }

    /// <summary>
    /// A string representation of the user agent
    /// </summary>
    /// <returns>string representation of the user agent</returns>
    public override string ToString()
    {
      StringBuilder sb = new StringBuilder(Family);
      string version = ToVersionString();

      if (!string.IsNullOrEmpty(version))
      {
        sb.Append(' ').Append(version);
      }

      return sb.ToString();
    }
    /// <summary>
    /// The version of the user agent as a single string
    /// </summary>
    /// <returns>the version as a string</returns>
    public string ToVersionString()
    {
      IEnumerable<string> versions = new string[] { Major, Minor, Patch }.Where(x => !string.IsNullOrEmpty(x));
      return string.Join(".", versions.ToArray());
    }
  }
}

File: go/uaparser/user_agent.go
package uaparser

import (
	"regexp"
	"strings"
)

type UserAgent struct {
	Family string
	Major  string
	Minor  string
	Patch  string
}

type UserAgentPattern struct {
	Regexp            *regexp.Regexp
	Regex             string
	FamilyReplacement string
	V1Replacement     string
	V2Replacement     string
}

func (uaPattern *UserAgentPattern) Match(line string, ua *UserAgent) {
	bytes := uaPattern.Regexp.FindStringSubmatch(line)
	if len(bytes) > 0 {
		groupCount := uaPattern.Regexp.NumSubexp()

		if len(uaPattern.FamilyReplacement) > 0 {
			if strings.Contains(uaPattern.FamilyReplacement, "$1") && groupCount >= 1 && len(bytes) >= 2 {
				ua.Family = strings.Replace(uaPattern.FamilyReplacement, "$1", bytes[1], 1)
			} else {
				ua.Family = uaPattern.FamilyReplacement
			}
		} else if groupCount >= 1 {
			ua.Family = bytes[1]
		}

		if len(uaPattern.V1Replacement) > 0 {
			ua.Major = uaPattern.V1Replacement
		} else if groupCount >= 2 {
			ua.Major = bytes[2]
		}

		if len(uaPattern.V2Replacement) > 0 {
			ua.Minor = uaPattern.V2Replacement
		} else if groupCount >= 3 {
			ua.Minor = bytes[3]
			if groupCount >= 4 {
				ua.Patch = bytes[4]
			}
		}
	}
}

func (ua *UserAgent) ToString() string {
	var str string
	if ua.Family != "" {
		str += ua.Family
	}
	version := ua.ToVersionString()
	if version != "" {
		str += " " + version
	}
	return str
}

func (ua *UserAgent) ToVersionString() string {
	var version string
	if ua.Major != "" {
		version += ua.Major
	}
	if ua.Minor != "" {
		version += "." + ua.Minor
	}
	if ua.Patch != "" {
		version += "." + ua.Patch
	}
	return version
}
